[libc++][test] Enhance `ADDITIONAL_COMPILE_FLAGS` to work with MSVC

StephanTLavavej opened the pull request:
Found while running libc++'s tests with MSVC's STL.

The `ADDITIONAL_COMPILE_FLAGS` machinery is very useful, but also very problematic for MSVC, as it speaks Martian and doesn't understand most of Clang's compiler options. (`-Dmeow` is mutually intelligible and that's about it.) We've been dealing with this by simply marking anything that uses `ADDITIONAL_COMPILE_FLAGS` as `FAIL` or `SKIPPED`, but that creates significant gaps in test coverage.

Fortunately, `ADDITIONAL_COMPILE_FLAGS` also supports "features", which can be slightly enhanced to send Clang-compatible and MSVC-compatible options to the right compilers.

* Add a new feature `gcc-style-warnings`.
  + The idea is that this should be defined for Apple Clang, vanilla Clang, or GCC, but not MSVC's `cl.exe`. Then this feature should be used when a test wants to compile with `-Wno-meow`. This is the only "ongoing maintenance" request here, which should be minimal and easy to fix if any occurrences are forgotten under `libcxx/test/std`.
  + In `libcxx/utils/libcxx/test/features.py`, I've implemented this by adding `_isAnyClangOrGCC`. I've also slightly refactored the logic (no functional changes) by adding `_isAnyClang` and `_isAnyGCC`, along the lines of @h-vetinari's suggestion.
* Also add `(gcc-style-warnings)` to an inactive `ADDITIONAL_COMPILE_FLAGS` in `libcxx/test/std/iterators/iterator.primitives/iterator.operations/advance.pass.cpp`, so it won't be forgotten.
* Add `ADDITIONAL_COMPILE_FLAGS(gcc-style-warnings): -Wno-undefined-inline` to `range.lazy.split/constraints.compile.pass.cpp`.
  + As the comment explains, this is compile-only, so the warning is irrelevant.
* Add `ADDITIONAL_COMPILE_FLAGS(cl-style-warnings)`.
  + The idea is that this should be defined for MSVC's `cl.exe` only, so it's defined as the antonym of `_isAnyClangOrGCC`.
  + Note that the existing `msvc` feature is completely different - it indicates the presence of the `_MSC_VER` compiler macro, i.e. any compiler pretending to be MSVC. https://github.com/llvm/llvm-project/blob/b85f1f9b182234ba366d78ae2174a149e44d08c1/libcxx/utils/libcxx/test/features.py#L19 https://github.com/llvm/llvm-project/blob/b85f1f9b182234ba366d78ae2174a149e44d08c1/libcxx/utils/libcxx/test/features.py#L117
  + @philnik777's naming idea `cl-style-warnings` makes this distinction clear - it indicates the actual `cl.exe` compiler driver.
    - Fun fact, nobody knows what CL means anymore. I asked our longest-tenured compiler devs (30+ years!) and their best guess was "Compile and Link". :joy_cat:
  + Because MSVC's `/wdNNNN` isn't self-explanatory, I'm commenting every occurrence with what the warning message is.
  + Some warnings being silenced are sign/truncation warnings that were intentional (e.g. caused by calling STL algorithms with types that cause the STL to perform mixed-sign comparisons), or that were relatively innocuous and would be invasive or verbose to physically fix. I tried to fix as much as I could in #74962, #74965, and previous PRs.
  + Other warnings being silenced are counterparts of Clang warnings being silenced (e.g. `-Wno-deprecated-volatile`).
  + "warning C4197: top-level `volatile` in cast is ignored" is emitted for `T();` when `T` is `volatile atomic<Whatever>`, which is innocuous.
  + "warning C4611: interaction between '`_setjmp`' and C++ object destruction is non-portable" is emitted in `setjmp` tests. I didn't see any concerning usage.
  + "warning C4583: '`X::cv_`': destructor is not implicitly called" was emitted in a test that explicitly destroys `x.cv_` so it's fine.
  + There's *no* expectation that as libc++ tests are added, that `ADDITIONAL_COMPILE_FLAGS(cl-style-warnings)` ought to be added accordingly. In fact, I'd prefer it if you didn't, allowing us to review such warnings as they appear. (Of course, if a test is copy-paste-modified, it would be reasonable to preserve any warning options that are unaffected by the modifications, that would never be a problem for us.)
* Add `ADDITIONAL_COMPILE_FLAGS(cl-style-warnings)` to silence a blizzard of x86 truncation warnings.

StephanTLavavej committed:
Use the verify-support feature.

StephanTLavavej committed:
Add a new feature any-clang-or-gcc.

StephanTLavavej committed:
Add ADDITIONAL_COMPILE_FLAGS(msvc-cl-exe).

StephanTLavavej committed:
Add ADDITIONAL_COMPILE_FLAGS(any-clang-or-gcc).

StephanTLavavej committed:
Add ADDITIONAL_COMPILE_FLAGS(msvc-cl-exe) to silence a blizzard of x86 truncation warnings.

StephanTLavavej committed:
Apply Python formatting from CI.

StephanTLavavej committed:
Properly implement `_isAnyClangOrGCC`.

StephanTLavavej committed:
Apply Python diff from CI.

cpplearner commented:
IMHO, `TEST_CLANG_DIAGNOSTIC_IGNORED`/`TEST_GCC_DIAGNOSTIC_IGNORED`/`TEST_MSVC_DIAGNOSTIC_IGNORED` (from [test_macros.h](https://github.com/llvm/llvm-project/blob/78623b079b3be841e96ce968ae5156fe26f6c565/libcxx/test/support/test_macros.h#L408-L432)) should be the preferred way to disable warnings.

philnik777 commented on libcxx/test/std/atomics/atomics.types.generic/trivially_copyable.verify.cpp:
```diff
@@ -15,7 +15,7 @@
 // errors while trying to instantiate the rest of std::atomic<T>.
 // We silence those to make the test more robust.
-// ADDITIONAL_COMPILE_FLAGS: -Xclang -verify-ignore-unexpected=error
+// ADDITIONAL_COMPILE_FLAGS(verify-support): -Xclang -verify-ignore-unexpected=error
```
I'm somewhat surprised that you're running any `.verify.cpp` tests. I was always under the impression that they are basically useless for anybody but libc++. How exactly are they useful to you?

StephanTLavavej replied:
I think we're skipping them - I added this during my initial round of trying to properly guard all of the compile flags, but I didn't check to see whether we actually need this. I'll check now.

StephanTLavavej replied:
Yep, great catch, we didn't need this! Reverted.

philnik777 reviewed (commented):
I think I would name the features `gcc-style-warnings` and `cl-style-warnings` to make it clear what the intended use here is. I don't think we want anybody to start using these flags for anything else.

StephanTLavavej committed:
Revert "Use the verify-support feature."

This reverts commit e37385902de4323ad6fcc18f4f4d4cfe582a5c02.

StephanTLavavej committed:
Rename `any-clang-or-gcc` => `gcc-style-warnings`, `msvc-cl-exe` => `cl-style-warnings`.

I've left `_isAnyClangOrGCC` unchanged, since that's what it's physically checking for.

StephanTLavavej commented:
I've reverted the `verify-support` changes, renamed `any-clang-or-gcc` => `gcc-style-warnings` and `msvc-cl-exe` => `cl-style-warnings`, and updated the PR description.

In the Python script, I've left `_isAnyClangOrGCC` unchanged, since that's what it's physically checking for. Let me know if you want that renamed too.

h-vetinari commented on libcxx/utils/libcxx/test/features.py:
```diff
@@ -13,7 +13,12 @@
 import sys
 
-_isClang = lambda cfg: "__clang__" in compilerMacros(cfg) and "__apple_build_version__" not in compilerMacros(cfg)
+_isAnyClangOrGCC = lambda cfg: "__clang__" in compilerMacros(
+    cfg
+) or "__GNUC__" in compilerMacros(cfg)
```
Could be deduplicated (both in terms of code and macro evaluations), e.g.:
```suggestion
_isAnyClangOrGCC = lambda cfg: _isClang(cfg) or _isAppleClang(cfg) or _isGCC(cfg)
```

(Obviously this would have to be added further down, after `_isGCC`.)

The potential for deduplication is now just about this PR though (though it certainly doesn't improve the situation); already the existing options could be deduplicated a bunch (which actually gets more attractive with the frankly horrible formatting that black chooses for the definition of `_isClang`):
```python
_isAnyClang = lambda cfg: "__clang__" in compilerMacros(cfg)
_isAppleClang = lambda cfg: "__apple_build_version__" in compilerMacros(cfg)
_isClang = lambda cfg: _isAnyClang(cfg) and not _isAppleClang(cfg)
_isGCC = lambda cfg: "__GNUC__" in compilerMacros(cfg) and not _isAnyClang(cfg)
_isAnyClangOrGCC = lambda cfg: _isAnyClang(cfg) or _isGCC(cfg)
```

StephanTLavavej replied:
Great idea, I've pushed a commit to do that. Note that I took a slightly different approach, introducing `_isAnyGCC` so that `_isAnyClangOrGCC`'s evaluation will be more efficient.

StephanTLavavej committed:
Refactor `_isMeow` logic for clarity.

ldionne commented on libcxx/utils/libcxx/test/features.py:
```diff
@@ -13,9 +13,12 @@
 import sys
 
-_isClang = lambda cfg: "__clang__" in compilerMacros(cfg) and "__apple_build_version__" not in compilerMacros(cfg)
+_isAnyClang = lambda cfg: "__clang__" in compilerMacros(cfg)
```
I would much prefer if we had a way of detecting `cl-style-warnings` here. I am waging a war against stale features in our test suite, and `cl-style-warnings` without any way to detect it looks exactly like a stale Lit feature.

ldionne reviewed (commented):
This makes sense to me and it seems like a reasonable support burden for us. I have a comment, but this looks pretty good to me.

> IMHO, `TEST_CLANG_DIAGNOSTIC_IGNORED`/`TEST_GCC_DIAGNOSTIC_IGNORED`/`TEST_MSVC_DIAGNOSTIC_IGNORED` (from [test_macros.h](https://github.com/llvm/llvm-project/blob/78623b079b3be841e96ce968ae5156fe26f6c565/libcxx/test/support/test_macros.h#L408-L432)) should be the preferred way to disable warnings.

@cpplearner  This is an interesting point. We could also decide to go all out and use the macros instead. TBH I am mostly neutral on this, but I'd be curious to hear what motivates your preference.

StephanTLavavej committed:
Add `(gcc-style-warnings)` to an inactive `ADDITIONAL_COMPILE_FLAGS` in iterator.operations/advance.pass.cpp.

StephanTLavavej committed:
Define `cl-style-warnings` when `not _isAnyClangOrGCC(cfg)`.

StephanTLavavej commented:
I think I can prepare an alternative PR to use the macros everywhere. It'll take a bit of time, but I've already done the hard work here of figuring out which files need to suppress which MSVC warnings.

StephanTLavavej commented:
I've created #75133 and would prefer to merge that PR, since I was able to locally suppress some warnings and remove some weirdness from `msvc_stdlib_force_include.h`. There are already a fair number of occurrences of using the macros for warning suppression, so it increases codebase consistency.

StephanTLavavej commented:
Closing this PR since I want to proceed with the macro approach.
